### PR TITLE
[#1747] Bugs default values with variables

### LINKED ISSUE #1747: Default values for the variables
### Product versie / Product version

latest

### Omschrijf het probleem / Describe the bug

Issues:
- Here there are initial values for the variables of different types: https://github.com/open-formulieren/open-forms/blob/master/src/openforms/forms/models/form_variable.py#L36. But, `boolean` should have `False` as in Formio, `float` should have `None`. 
- The file component is of type object, but having `{}` as default value breaks it.
- When checking what data we should return here https://github.com/open-formulieren/open-forms/blob/master/src/openforms/submissions/models/submission_value_variable.py#L36 we check that the value is not an empty string. We should check against the default value/empty value.
- If a variable has a default value, even if the form field is hidden it ends up in the final form data. => Note: if the logic makes it visible and then hidden again, also without variables this value ends up in the data. So this is a separate bug (https://github.com/open-formulieren/open-forms/issues/1755).

### PR DESCRIPTION
**Pull request by SilviaAmAm**
Fixes #1747 

**Changes**

- Before, all variables related to a step were included in the step data with their default value at the beginning of the step.
Now, if a variable is not in the configuration (i.e. its hidden) then it is not added to the data.
- The component without an explicit default value have a corresponding component variable with `initial_value=None`.
- The function `bulk_create_or_update_from_data` deletes any saved variables which are no longer present in the data.
- User defined variables are initialised (saved) at the beginning of filling out the form (after prefills).

### CODE REVIEW

File: src/openforms/forms/models/form_variable.py [hunk 1]
```diff
@@ -212,7 +202,7 @@ def get_initial_value(self):
         if self.source == FormVariableSources.static:
             return STATIC_INITIAL_VALUES[self.initial_value]()
         else:
-            return self.initial_value or INITIAL_VALUES[self.data_type]
+            return self.initial_value or None
```
Comment (sergei-maertens) [inline src/openforms/forms/models/form_variable.py:206]: How does this behave with an empty string is default value? I'm not sure how Formio saves that in the form builder.

File: src/openforms/forms/models/form_variable.py [hunk 2]
```diff
@@ -97,6 +86,7 @@ def create_for_formstep(self, form_step: "FormStep") -> List["FormVariable"]:
                     is_sensitive_data=component.get("isSensitiveData", False),
                     source=FormVariableSources.component,
                     data_type=get_component_datatype(component),
+                    initial_value=component.get("defaultValue"),
```
Comment (SilviaAmAm) [inline src/openforms/forms/models/form_variable.py:89]: Make a `get_default_value` in utils for formio documenting that Formio has the empty value and default value concepts + links to github

File: src/openforms/js/components/admin/form_design/form-creation-form.js [hunk 3]
```diff
@@ -360,7 +360,7 @@ function reducer(draft, action) {
 
       // Issue #1729 - Workaround for bug in FormIO
       if (mutationType === 'changed' && !schema.multiple && Array.isArray(schema.defaultValue)) {
-        schema.defaultValue = getEmptyValue(schema);
+        schema.defaultValue = null;
```
Comment (SilviaAmAm) [inline src/openforms/js/components/admin/form_design/form-creation-form.js:363]: comment that formio will fallback to the emptyvalue of the component.

File: src/openforms/submissions/api/viewsets.py [hunk 4]
```diff
@@ -119,6 +120,7 @@ def perform_create(self, serializer):
         conf = GlobalConfiguration.get_solo()
         if conf.enable_form_variables:
             prefill_variables(serializer.instance)
+            initialise_variables_unrelated_to_a_step(serializer.instance)
```
Comment (SilviaAmAm) [inline src/openforms/submissions/api/viewsets.py:123]: When resuming a form we will have to refresh the static variables?
Maybe we shouldn't persist static variables to databases?

File: src/openforms/submissions/models/submission_value_variable.py [hunk 5]
```diff
@@ -18,41 +18,82 @@
...
+
+    @property
+    def saved_variables(self):
+        return {
+            variable_key: variable
+            for variable_key, variable in self.variables.items()
+            if variable.pk
+        }
 
     def get_variable(self, key: str) -> Optional["SubmissionValueVariable"]:
```
Comment (SilviaAmAm) [inline src/openforms/submissions/models/submission_value_variable.py:41]: Cannot return none?

File: src/openforms/submissions/models/submission_value_variable.py [hunk 6]
```diff
@@ -18,41 +18,82 @@
...
 class SubmissionValueVariablesState:
-    variables: Dict[str, "SubmissionValueVariable"]
+    submission: "Submission"
+    _variables: Dict[str, "SubmissionValueVariable"] = None
+
+    def __init__(self, submission: "Submission"):
+        self.submission = submission
+
+    @property
+    def variables(self):
```
Comment (SilviaAmAm) [inline src/openforms/submissions/models/submission_value_variable.py:28]: Add type hints

File: src/openforms/submissions/models/submission_value_variable.py [hunk 7]
```diff
@@ -115,12 +151,17 @@ def bulk_create_or_update_from_data(
...
         variables_to_update = []
+        variables_ids_to_delete = []
         for key, variable in submission_variables.items():
             try:
                 variable.value = glom(data, key)
             except PathAccessError:
                 if update_missing_variables:
-                    variable.value = variable.form_variable.get_initial_value()
+                    if variable.pk:
+                        variables_ids_to_delete.append(variable.id)
```
Comment (SilviaAmAm) [inline src/openforms/submissions/models/submission_value_variable.py:161]: Maybe just remove the deleted variables from memory? Save some queries

File: src/openforms/js/components/admin/form_design/variables/utils.js [hunk 8]
```diff
@@ -64,7 +64,7 @@ const updateFormVariables = (
           prefillPlugin: newComponent.prefill?.plugin || '',
           prefillAttribute: newComponent.prefill?.attribute || '',
           isSensitiveData: newComponent.isSensitiveData,
-          initialValue: newComponent.defaultValue || '',
+          initialValue: getDefaultValue(newComponent),
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openforms/js/components/admin/form_design/variables/utils.js:67]: Created an issues for this: https://github.com/open-formulieren/open-forms/issues/1760

File: src/openforms/submissions/api/viewsets.py (hunk 4 above)
Comment (SilviaAmAm) [reply src/openforms/submissions/api/viewsets.py:123]: Created issue: https://github.com/open-formulieren/open-forms/issues/1761

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**